Don't work with Dokku 0.4

I believe that plugin system changed in v0.4. Now you can list all installed plugins by dokku command. There's an error in that listing:

```
# dokku plugin
  00_dokku-standard    0.4.0 enabled    dokku core standard plugin
  20_events            0.4.0 enabled    dokku core events logging plugin
  app-predeploy-tasks  0.1.0 enabled    Run arbitrary app tasks prior to dokku's deploy phase
  apps                 0.4.0 enabled    dokku core apps plugin
  backup               0.4.0 enabled    dokku core backup plugin
  build-env            0.4.0 enabled    dokku core build-env plugin
  certs                0.4.0 enabled    dokku core certificate management plugin
  checks               0.4.0 enabled    dokku core checks plugin
  common               0.4.0 enabled    dokku core common plugin
  config               0.4.0 enabled    dokku core config plugin
  docker-options       0.4.0 enabled    dokku core docker-options plugin
  domains              0.4.0 enabled    dokku core domains plugin
  enter                0.4.0 enabled    dokku core enter plugin
  git                  0.4.0 enabled    dokku core git plugin
cat: /var/lib/dokku/plugins/available/logging-supervisord/plugin.toml: No such file or directory
panic: interface conversion: interface is nil, not map[string]interface {}

goroutine 1 [running]:
runtime.panic(0x5ffd60, 0xc2100520c0)
	/usr/lib/go/src/pkg/runtime/panic.c:266 +0xb6
main.TomlGet(0xc21000a030, 0x2, 0x2)
	/tmp/tmp/plugn/plugn.go:30 +0x1d1
github.com/progrium/go-basher.(*Context).HandleFuncs(0xc21000f320, 0xc21000a000, 0x5, 0x5, 0x0)
	/home/vagrant/gocode/src/github.com/progrium/go-basher/basher.go:182 +0x2ac
github.com/progrium/go-basher.Application(0xc21001edb0, 0x7fa13ec1df00, 0x4, 0x4, 0x6aeb20, ...)
	/home/vagrant/gocode/src/github.com/progrium/go-basher/basher.go:64 +0x4d2
main.main()
	/tmp/tmp/plugn/plugn.go:98 +0x431

goroutine 3 [runnable]:
os/signal.loop()
	/usr/lib/go/src/pkg/os/signal/signal_unix.go:19
created by os/signal.init·1
	/usr/lib/go/src/pkg/os/signal/signal_unix.go:27 +0x31
cat: /var/lib/dokku/plugins/available/logging-supervisord/plugin.toml: No such file or directory
panic: interface conversion: interface is nil, not map[string]interface {}

goroutine 1 [running]:
runtime.panic(0x5ffd60, 0xc2100520c0)
	/usr/lib/go/src/pkg/runtime/panic.c:266 +0xb6
main.TomlGet(0xc21000a030, 0x2, 0x2)
	/tmp/tmp/plugn/plugn.go:30 +0x1d1
github.com/progrium/go-basher.(*Context).HandleFuncs(0xc21000f320, 0xc21000a000, 0x5, 0x5, 0x0)
	/home/vagrant/gocode/src/github.com/progrium/go-basher/basher.go:182 +0x2ac
github.com/progrium/go-basher.Application(0xc21001edb0, 0x7fa1b8c4af00, 0x4, 0x4, 0x6aeb20, ...)
	/home/vagrant/gocode/src/github.com/progrium/go-basher/basher.go:64 +0x4d2
main.main()
	/tmp/tmp/plugn/plugn.go:98 +0x431

goroutine 3 [runnable]:
os/signal.loop()
	/usr/lib/go/src/pkg/os/signal/signal_unix.go:19
created by os/signal.init·1
	/usr/lib/go/src/pkg/os/signal/signal_unix.go:27 +0x31
  logging-supervisord        enabled
  logs                 0.4.0 enabled    dokku core logs plugin
  nginx-vhosts         0.4.0 enabled    dokku core nginx-vhosts plugin
  plugin               0.4.0 enabled    dokku core plugin plugin
  postgres             1.0.0 enabled    dokku postgres service plugin
  ps                   0.4.0 enabled    dokku core ps plugin
  shell                0.4.0 enabled    dokku core shell plugin
  tags                 0.4.0 enabled    dokku core tags plugin
  tar                  0.4.0 enabled    dokku core tar plugin
```

Maybe we can take an inspiration what to update from `app-predeploy-tasks` plugin, which as you can see works.

--- a/Makefile
+++ b/Makefile
@@ -1,0 +1,38 @@
+.PHONY: test test-default test-run-filter
+
+PROCFILE_PATH := "/tmp/Procfile"
+SCALE_FILE_PATH := "/tmp/SCALE"
+
+test: test-default test-run-filter
+
+test-default:
+	@$(MAKE) setup-procfile setup-scale-file
+	@echo testing defaults
+	@$(QUIET) [[ `bash lib/procfile-to-supervisord $(PROCFILE_PATH) $(SCALE_FILE_PATH) | grep "\[progra" | wc -l` -eq 4 ]] || exit 1
+	@echo test succeeded
+	@$(MAKE) clean
+
+test-run-filter:
+	@$(MAKE) setup-procfile setup-scale-file
+	@echo testing DOKKU_SUPERVISOR_RUN
+	@$(QUIET) [[ `unset DOKKU_SUPERVISOR_RUN; DOKKU_SUPERVISOR_RUN="web lowlatencyworker" bash lib/procfile-to-supervisord $(PROCFILE_PATH) $(SCALE_FILE_PATH) | grep "\[progra" | wc -l` -eq 2 ]] || exit 1
+	@echo test succeeded
+	@$(MAKE) clean
+
+setup-procfile:
+	@echo "web: run webserver" > $(PROCFILE_PATH)
+	@echo "assets: build assets" >> $(PROCFILE_PATH)
+	@echo "db-migrate: migrate db" >> $(PROCFILE_PATH)
+	@echo "urgentworker: run urgentworker" >> $(PROCFILE_PATH)
+	@echo "worker: run worker" >> $(PROCFILE_PATH)
+	@echo "lowlatencyworker: run lowlatencyworker" >> $(PROCFILE_PATH)
+
+setup-scale-file:
+	@echo "web=1" > $(SCALE_FILE_PATH)
+	@echo "worker=1" >> $(SCALE_FILE_PATH)
+	@echo "urgentworker=1" >> $(SCALE_FILE_PATH)
+	@echo "assets=0" >> $(SCALE_FILE_PATH)
+	@echo "db-migrate=0" >> $(SCALE_FILE_PATH)
+
+clean:
+	@$(QUIET) rm -f $(PROCFILE_PATH) $(SCALE_FILE_PATH)

--- a/README.md
+++ b/README.md
@@ -10,10 +10,6 @@ This plugin uses the `docker-args` hook to inject the data volume argument. As s
 ## Installation
 
 ```sh
-# Create the directory to house the log files:
-sudo mkdir -p /var/log/dokku
-sudo chown dokku:dokku /var/log/dokku
-
 # Install the plugin:
 # dokku 0.3.x
 git clone https://github.com/sehrope/dokku-logging-supervisord.git /var/lib/dokku/plugins/logging-supervisord

--- a/commands
+++ b/commands
@@ -45,8 +45,11 @@ case "$1" in
     ;;
   help)
     cat && cat<<EOF
-    scale <app> TYPE1=NUM1 [TYPE2=NUM2 ...]         Scale an app
+    scale <app> TYPE1=NUM1 [TYPE2=NUM2 ...], Scale an app
 EOF
   ;;
+  *)
+    exit $DOKKU_NOT_IMPLEMENTED_EXIT
+    ;;
 
 esac

--- a/commands
+++ b/commands
@@ -39,7 +39,7 @@ case "$1" in
 
     echo "Scaling app $APP:"
     cat -n "$REPO/SCALE"
-    copy_to_container "$SCALEFILE" /SCALE
+    copy_to_container "$SCALEFILE" /app/SCALE
 
     dokku deploy "$APP"
     ;;

--- a/docker-args-deploy
+++ b/docker-args-deploy
@@ -1,0 +1,1 @@
+docker-args

--- a/docker-args-run
+++ b/docker-args-run
@@ -1,0 +1,1 @@
+docker-args

--- a/lib/procfile-to-supervisord
+++ b/lib/procfile-to-supervisord
@@ -53,24 +53,39 @@ gen_supervisor_conf
 
 while read line || [ -n "$line" ]
 do
-  if [ ! -n "$line" ]; then
-    continue
-  fi
-
+  [[ -z "$line" ]] && continue
+  RUN_PROC=false
   NAME=${line%%:*}
   COMMAND=${line#*: }
   NUM_PROCS=1
 
-  if [ -f "$SCALEFILE" ]; then
-    SCALE=$(grep "$NAME=" "$SCALEFILE")
-    NUM_PROCS=${SCALE#*=}
-    if [ ! "$NUM_PROCS" -ge 1 ]; then
-      echo "Invalid scale line for process $NAME: $SCALE" 1>&2
-      exit 1
+  if [[ -f "$SCALEFILE" ]]; then
+    SCALE=$(egrep "^$NAME=" "$SCALEFILE")
+    if [[ -n "$SCALE" ]]; then
+      NUM_PROCS=${SCALE#*=}
+      if [[ "$NUM_PROCS" -eq 0 ]]; then
+        echo "=====> Not running $NAME as scale is zero" >&2
+        continue
+      fi
     fi
   fi
 
-  # Add a blank line in between processes:
-  echo ""
-  gen_supervisor_proc_conf "$NAME" "$COMMAND" "$NUM_PROCS"
+  if [[ -n "$DOKKU_SUPERVISOR_RUN" ]]; then
+    for PROC in $DOKKU_SUPERVISOR_RUN; do
+      if [[ "$PROC" = "$NAME" ]]; then
+        RUN_PROC=true
+        break
+      fi
+    done
+  else
+    RUN_PROC=true
+  fi
+
+  if [[ "$NUM_PROCS" -ge 1 ]]; then
+    if [[ "$RUN_PROC" = "true" ]];then
+      # Add a blank line in between processes:
+      echo ""
+      gen_supervisor_proc_conf "$NAME" "$COMMAND" "$NUM_PROCS"
+    fi
+  fi
 done < "$PROCFILE"

--- a/lib/procfile-to-supervisord
+++ b/lib/procfile-to-supervisord
@@ -51,7 +51,8 @@ EOF
 
 gen_supervisor_conf
 
-cat "$PROCFILE" | while read line; do
+while read line || [ -n "$line" ]
+do
   if [ ! -n "$line" ]; then
     continue
   fi
@@ -72,4 +73,4 @@ cat "$PROCFILE" | while read line; do
   # Add a blank line in between processes:
   echo ""
   gen_supervisor_proc_conf "$NAME" "$COMMAND" "$NUM_PROCS"
-done
+done < "$PROCFILE"

--- a/lib/procfile-to-supervisord
+++ b/lib/procfile-to-supervisord
@@ -38,12 +38,12 @@ fi
 
 cat << EOF
 [program:${NAME}]
-command=/exec sh -c '${COMMAND}'
+command=/exec ${COMMAND}
 process_name:${PROCESS_NAME}
 numprocs=${NUM_PROCS}
 autostart=true
 autorestart=true
-stopsignal=QUIT
+stopsignal=TERM
 redirect_stderr=true
 stdout_logfile=/var/log/app/%(program_name)s.%(process_num)02d.log
 EOF

--- a/lib/start
+++ b/lib/start
@@ -6,7 +6,7 @@ hash -r
 cd $HOME
 
 # Generate supervisord.conf:
-bash /build/procfile-to-supervisord /app/Procfile /app/SCALE > supervisord.conf
+bash /usr/local/bin/procfile-to-supervisord /app/Procfile /app/SCALE > supervisord.conf
 
 # Create /var/log/app directory
 mkdir -p /var/log/app

--- a/lib/start
+++ b/lib/start
@@ -24,13 +24,14 @@ mkdir -p /var/log/app
 # Get buildstep user if it exists and chown log dir
 user_name=$(stat -c %U /app)
 user_group=$(stat -c %G /app)
+mount_list="/var/log/app /mnt/whoosh_index /mnt/simsrv"
 if [[ -n "$user_name" ]] && [[ -n "$user_group" ]];then
-    chown -R ${user_name}:${user_group} /var/log/app
-    chmod -R g+rw /var/log/app
-    if [ -d "/mnt/whoosh_index" ];then
-        chown -R ${user_name}:${user_group} /mnt/whoosh_index
-        chmod -R g+rw /mnt/whoosh_index
-    fi
+    for mount in $mount_list; do
+        if [ -d "$mount" ];then
+            chown -R ${user_name}:${user_group} "$mount"
+            chmod -R g+rw "$mount"
+        fi
+    done
 fi
 
 # Display the generated supervisord.conf:

--- a/lib/start
+++ b/lib/start
@@ -27,6 +27,10 @@ user_group=$(stat -c %G /app)
 if [[ -n "$user_name" ]] && [[ -n "$user_group" ]];then
     chown -R ${user_name}:${user_group} /var/log/app
     chmod -R g+rw /var/log/app
+    if [ -d "/mnt/whoosh_index" ];then
+        chown -R ${user_name}:${user_group} /mnt/whoosh_index
+        chmod -R g+rw /mnt/whoosh_index
+    fi
 fi
 
 # Display the generated supervisord.conf:

--- a/lib/start
+++ b/lib/start
@@ -24,14 +24,9 @@ mkdir -p /var/log/app
 # Change ownership of /var/log/app to non-root container user
 APP_USER=$(stat -c %U /app)
 APP_GROUP=$(stat -c %G /app)
-mount_list="/var/log/app /mnt/whoosh_index /mnt/simsrv"
 if [[ -n "$APP_USER" ]] && [[ -n "$APP_GROUP" ]]; then
-  for mount in $mount_list; do
-    if [[ -d "$mount" ]];then
-      chown -R "${APP_USER}:${APP_GROUP}" "$mount"
-      chmod -R g+rw "$mount"
-    fi
-  done
+  chown -R "${APP_USER}:${APP_GROUP}" /var/log/app
+  chmod -R g+rw /var/log/app
 fi
 
 # Display the generated supervisord.conf:

--- a/lib/start
+++ b/lib/start
@@ -21,6 +21,14 @@ bash /usr/local/bin/procfile-to-supervisord /app/Procfile /app/SCALE > superviso
 # Create /var/log/app directory
 mkdir -p /var/log/app
 
+# Get buildstep user if it exists and chown log dir
+user_name=$(stat -c %U /app)
+user_group=$(stat -c %G /app)
+if [[ -n "$user_name" ]] && [[ -n "$user_group" ]];then
+    chown -R ${user_name}:${user_group} /var/log/app
+    chmod -R g+rw /var/log/app
+fi
+
 # Display the generated supervisord.conf:
 echo "Generated supervisord.conf:"
 cat -n supervisord.conf

--- a/lib/start
+++ b/lib/start
@@ -5,6 +5,16 @@ export HOME=/app
 hash -r
 cd $HOME
 
+# source in app env for things like DOKKU_SUPERVISOR_RUN
+if [[ -d /app/.profile.d ]]; then
+  for file in /app/.profile.d/*.sh; do
+    source $file
+  done
+fi
+# need to unset PYTHON vars to use apt installed supervisord
+unset PYTHONHOME
+unset PYTHONPATH
+
 # Generate supervisord.conf:
 bash /usr/local/bin/procfile-to-supervisord /app/Procfile /app/SCALE > supervisord.conf
 

--- a/plugin.toml
+++ b/plugin.toml
@@ -1,0 +1,6 @@
+[plugin]
+name = "dokku-logging-supervisord"
+description = "A plugin for dokku that injects supervisord"
+version = "0.1.0"
+website_url = "https://github.com/expa/dokku-logging-supervisord"
+[plugin.config]

--- a/plugin.toml
+++ b/plugin.toml
@@ -2,5 +2,5 @@
 name = "dokku-logging-supervisord"
 description = "A plugin for dokku that injects supervisord"
 version = "0.1.0"
-website_url = "https://github.com/expa/dokku-logging-supervisord"
+website_url = "https://github.com/sehrope/dokku-logging-supervisord"
 [plugin.config]

--- a/post-release
+++ b/post-release
@@ -14,7 +14,7 @@ if [ $(docker wait $id) -ne 0 ]; then
   exit 0
 fi
 
-copy_to_container "$PLUGIN_DIR/lib/procfile-to-supervisord" /build/procfile-to-supervisord
+copy_to_container "$PLUGIN_DIR/lib/procfile-to-supervisord" /usr/local/bin/procfile-to-supervisord
 if [ -f "$SCALE_FILE" ]; then
   echo "Found SCALE file: $SCALE_FILE"
   copy_to_container "$SCALE_FILE" /app/SCALE
@@ -26,12 +26,12 @@ if docker run -a stdout -i $IMAGE /usr/bin/dpkg -s supervisor > /dev/null; then
   echo "supervisor is already installed (skipping apt-get update/install)"
 else
   echo "supervisor is not installed (will install via apt-get)"
-  id=$(docker run -i -a stdin $IMAGE /bin/bash -c "apt-get update && apt-get install -y supervisor && apt-get clean")
-  test $(docker wait $id) -eq 0
-  docker commit $id $IMAGE > /dev/null
+  run_in_container "apt-get update && apt-get install -y supervisor && apt-get clean"
 fi
 
-# Replace /start with our custom runner:
-id=$(cat "$PLUGIN_DIR/lib/runner" | docker run -i -a stdin $IMAGE /bin/bash -c "cat > /start")
-test $(docker wait $id) -eq 0
-docker commit $id $IMAGE > /dev/null
+# Unlink /exec from /start
+# Original builstep uses `ln -nsf /start /exec` that forks supervisor repeatedly
+run_in_container "if [[ -L /exec ]];then rm -f /exec && cp -f /start /exec; fi"
+
+# Replace /start with our custom start program:
+copy_to_container "$PLUGIN_DIR/lib/start" /start

--- a/post-release
+++ b/post-release
@@ -1,9 +1,12 @@
 #!/bin/bash
 set -eo pipefail; [[ $DOKKU_TRACE ]] && set -x
+source "$PLUGIN_CORE_AVAILABLE_PATH/common/functions" || true
 
 APP="$1"
 IMAGE="dokku/$1"
 SCALE_FILE="$DOKKU_ROOT/$APP/SCALE"
+
+[[ -n "$DOKKU_API_VERSION" ]] && IMAGE=$(get_app_image_name $APP)
 
 PLUGIN_DIR=$(dirname $0)
 . "$PLUGIN_DIR/lib/helpers"

--- a/post-release-buildpack
+++ b/post-release-buildpack
@@ -1,0 +1,1 @@
+post-release

